### ISSUE TITLE
`no_std` support

### ISSUE THREAD

**Issue by jscatena88** [rusticata/rusticata-macros#6]
I am looking at using one of the upstream dependents of this library on a `no_std` target. Is there any appetite to support that use case? The following git patch enables allows no-std builds when the package is brought in with `default-features = false`. I don't believe it would be a breaking change unless a package was already declaring `default-features = flase` for some reason and depends on functionality I ended up placing behind the `std` feature flag.

```
Subject: [PATCH] no_std support added

---
 Cargo.toml    |  6 +++++-
 src/debug.rs  | 12 ++++++++----
 src/lib.rs    |  2 ++
 src/traits.rs |  2 ++
 4 files changed, 17 insertions(+), 5 deletions(-)

diff --git a/Cargo.toml b/Cargo.toml
index cb377fc..8bd6131 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,4 +12,8 @@ categories = ["parsing"]
 readme = "README.md"
 
 [dependencies]
-nom = { version = "7.0", default-features = false, features = ["std"] }
+nom = { version = "7.0", default-features = false, features = ["alloc"] }
+
+[features]
+default = ["std"]
+std = ["nom/std"]
diff --git a/src/debug.rs b/src/debug.rs
index ec8977b..74e80cb 100644
--- a/src/debug.rs
+++ b/src/debug.rs
@@ -1,11 +1,15 @@
 //! Helper functions and structures for debugging purpose
 
-use nom::combinator::{map, peek, rest};
-use nom::HexDisplay;
-use nom::IResult;
-use std::fmt;
+use alloc::{format, vec::Vec};
+use core::fmt;
+#[cfg(feature = "std")]
+use nom::{
+    combinator::{map, peek, rest},
+    HexDisplay, IResult,
+};
 
 /// Dump the remaining bytes to stderr, formatted as hex
+#[cfg(feature = "std")]
 pub fn dbg_dmp_rest(i: &[u8]) -> IResult<&[u8], ()> {
     map(peek(rest), |r: &[u8]| eprintln!("\n{}\n", r.to_hex(16)))(i)
 }
diff --git a/src/lib.rs b/src/lib.rs
index 8d23249..aa0def1 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -30,7 +30,9 @@
     unused_import_braces,
     unused_qualifications
 )]
+#![cfg_attr(not(feature = "std"), no_std)]
 
+extern crate alloc;
 pub mod combinator;
 pub mod debug;
 pub use macros::*;
diff --git a/src/traits.rs b/src/traits.rs
index 38f1402..7d041b1 100644
--- a/src/traits.rs
+++ b/src/traits.rs
@@ -1,5 +1,7 @@
 //! Common traits
 
+use alloc::vec::Vec;
+
 /// Common trait for structures serialization
 pub trait Serialize<O = Vec<u8>> {
     /// Type of serialization error
-- 
```

These changes have unblocked our use case for now but I would love to contribute it back upstream if there is interest. I can obviously prepare this as a proper pull request if that is the case.

Thanks,
Jason

**Comment by chifflier** [comment]
Hi @jscatena88 ,
adding support for `no_std` would definitely be a good feature for this crate. Would you mind resubmitting this as a pull request, so this can be merged with usual git workflow?
At first sight, the patch looks fine (maybe just add a comment to function `dbg_dmp_rest` to indicate that it requires `std`).

Thanks!

**Comment by chifflier** [comment]
Closing, merged in #7
